extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/cache/DefaultBasicAuthorizerCacheManager.java

```diff
@@ -0,0 +1,236 @@
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class DefaultBasicAuthorizerCacheManager implements BasicAuthorizerCacheManager
+{
+  private static final EmittingLogger LOG = new EmittingLogger(DefaultBasicAuthorizerCacheManager.class);
+
+  private final ConcurrentHashMap<String, Map<String, BasicAuthorizerUser>> cachedUserMaps;
+  private final ConcurrentHashMap<String, Map<String, BasicAuthorizerRole>> cachedRoleMaps;
+
+  private final Set<String> authorizerPrefixes;
+  private final Injector injector;
+  private final ObjectMapper objectMapper;
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+  private final DruidLeaderClient druidLeaderClient;
+  private final BasicAuthCommonCacheConfig commonCacheConfig;
+
+  private volatile ScheduledExecutorService exec;
+
+  @Inject
+  public DefaultBasicAuthorizerCacheManager(
+      Injector injector,
+      BasicAuthCommonCacheConfig commonCacheConfig,
+      @Smile ObjectMapper objectMapper,
+      @Coordinator DruidLeaderClient druidLeaderClient
+  )
+  {
+    this.injector = injector;
+    this.commonCacheConfig = commonCacheConfig;
+    this.objectMapper = objectMapper;
+    this.cachedUserMaps = new ConcurrentHashMap<>();
+    this.cachedRoleMaps = new ConcurrentHashMap<>();
+    this.authorizerPrefixes = new HashSet<>();
+    this.druidLeaderClient = druidLeaderClient;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+
+    LOG.info("Starting DefaultBasicAuthorizerCacheManager.");
+
+    try {
+      initUserMaps();
+
+      this.exec = Execs.scheduledSingleThreaded("BasicAuthorizerCacheManager-Exec--%d");
+
+      ScheduledExecutors.scheduleWithFixedDelay(
+          exec,
+          new Duration(commonCacheConfig.getPollingPeriod()),
+          new Duration(commonCacheConfig.getPollingPeriod()),
+          () -> {
+            try {
+              LOG.debug("Scheduled cache poll is running");
+              for (String authorizerPrefix : authorizerPrefixes) {
+                UserAndRoleMap userAndRoleMap = fetchUserAndRoleMapFromCoordinator(authorizerPrefix, false);
+                if (userAndRoleMap != null) {
+                  cachedUserMaps.put(authorizerPrefix, userAndRoleMap.getUserMap());
+                  cachedRoleMaps.put(authorizerPrefix, userAndRoleMap.getRoleMap());
+                }
+              }
+              LOG.debug("Scheduled cache poll is done");
+
+              long randomDelay = ThreadLocalRandom.current().nextLong(0, commonCacheConfig.getMaxRandomDelay());
+              LOG.debug("Inserting random polling delay of [%s] ms", randomDelay);
+              Thread.sleep(randomDelay);
```

Comment:
Better sleep at the beginning because every node will try to sync simultaneously at first if sleep after polling.

Reply:
Moved sleep